tests/unit/test_backups.py

```diff
@@ -0,0 +1,463 @@
+        ]:
+            self.charm.unit.status = ActiveStatus()
+            _render_pgbackrest_conf_file.reset_mock()
+            _create_bucket_if_not_exists.reset_mock()
+            with self.harness.hooks_disabled():
+                self.harness.update_relation_data(
+                    self.peer_rel_id,
+                    self.charm.app.name,
+                    {"cluster_initialised": "True"},
+                )
+            _render_pgbackrest_conf_file.return_value = True
+            _create_bucket_if_not_exists.side_effect = error
+            self.charm.backup.s3_client.on.credentials_changed.emit(
+                relation=self.harness.model.get_relation(S3_PARAMETERS_RELATION, self.s3_rel_id)
+            )
+            _render_pgbackrest_conf_file.assert_called_once()
+            _create_bucket_if_not_exists.assert_called_once()
+            self.assertIsInstance(self.charm.unit.status, BlockedStatus)
+            self.assertEqual(
+                self.charm.unit.status.message, FAILED_TO_ACCESS_CREATE_BUCKET_ERROR_MESSAGE
+            )
+            _can_use_s3_repository.assert_not_called()
+            _initialise_stanza.assert_not_called()
+            _start_stop_pgbackrest_service.assert_not_called()
+
+        # Test when it's not possible to use the S3 repository due to backups from another cluster.
+        _create_bucket_if_not_exists.reset_mock()
+        _create_bucket_if_not_exists.side_effect = None
+        _can_use_s3_repository.return_value = (False, "fake validation message")
+        self.charm.backup.s3_client.on.credentials_changed.emit(
+            relation=self.harness.model.get_relation(S3_PARAMETERS_RELATION, self.s3_rel_id)
+        )
+        self.assertIsInstance(self.charm.unit.status, BlockedStatus)
+        self.assertEqual(self.charm.unit.status.message, "fake validation message")
+        _create_bucket_if_not_exists.assert_called_once()
+        _can_use_s3_repository.assert_called_once()
+        _initialise_stanza.assert_not_called()
+        _start_stop_pgbackrest_service.assert_not_called()
+
+        # Test when the stanza can be initialised and the pgBackRest service can start.
+        _can_use_s3_repository.reset_mock()
+        _can_use_s3_repository.return_value = (True, None)
+        self.charm.backup.s3_client.on.credentials_changed.emit(
+            relation=self.harness.model.get_relation(S3_PARAMETERS_RELATION, self.s3_rel_id)
+        )
+        _can_use_s3_repository.assert_called_once()
+        _initialise_stanza.assert_called_once()
+        _start_stop_pgbackrest_service.assert_called_once()
+
+    @patch("charm.PostgreSQLBackups._change_connectivity_to_database")
+    @patch("charm.PostgreSQLBackups._list_backups")
+    @patch("charm.PostgreSQLBackups._execute_command")
+    @patch("charm.PostgresqlOperatorCharm.is_primary", new_callable=PropertyMock)
+    @patch("charm.PostgreSQLBackups._upload_content_to_s3")
+    @patch("datetime.datetime")
```

Comment:
Does patching `charm.PostgreSQLBackups.datetime` not work?

Reply:
Unfortunately, it doesn't work. I also tested some other approaches, like `charm.datetime`, but none of them worked. Then I added the approach from this PR, that was copied from MySQL VM charm.

Reply:
Cool, let me try it here. Thanks for the suggestion Dragomir.